functionary/vllm_monkey_patch/async_llm_engine.py

```diff
@@ -0,0 +1,724 @@
+                tool_or_func["name"]
+                for tool_or_func in self.tools_or_functions[request_id]
+            ]
+            if self.prompt_templates[request_id].version not in ["v1"]:
+                options.append("all")
+        elif gen_state["stage"] == "parameter-name":
+            # Get the list of parameters for the curr_fn_name
+            curr_fn_name = gen_state["func_name"]
+            for tool_or_func in self.tools_or_functions[request_id]:
+                if tool_or_func["name"] == curr_fn_name:
+                    options = list(tool_or_func["parameters"]["properties"].keys())
+                    break
+        # No grammar sampling needed if gen_state not in "function" or "parameter-name"
+        # stages. Just return the model_sampled_token_id
+        else:
+            model_sampled_token = self.tokenizer.decode([model_sampled_token_id])
+            return model_sampled_token_id, model_sampled_token
+
+        # Loop through the list of token ids sorted in descending order
+        # Form a mask made up of booleans where the index of the mask ==
+        # index of function/parameter name in func_options/parameter_options.
+        # The element is True if the sampled_token helps in forming the
+        # function/parameter name. Else, False.
+        for i, sampled_token_ind in enumerate(delta_token_ids):
+            sampled_token = self.tokenizer.decode(
+                [sampled_token_ind], add_special_tokens=False
+            )
+
+            if gen_state["stage"] == "function":
+                # Form the function name with the current sampled token id
+                new_curr_tokens_id = gen_state["curr_tokens"] + [sampled_token_ind]
+                new_curr_tokens = self.tokenizer.decode(new_curr_tokens_id)
+
+                options_mask = [
+                    True if option.startswith(new_curr_tokens.lstrip(" ")) else False
+                    for option in options
+                ]
+
+                # - In case of two fns having common prefixes (e.g.: get_weather and
+                # get_weather_and_time), we need to iterate one more time to see if
+                # the next sampled token is in fn_param_sep_token to know if the shorter
+                # or longer function name is preferred by the model.
+                # - Reject the whitespace (" ") and empty ("") tokens
+                if (
+                    prompt_template.fn_param_sep_token.startswith(sampled_token)
+                    or any(options_mask)
+                ) and sampled_token.strip(" ") != "":
+                    return sampled_token_ind, sampled_token
+            else:
+                # Mask away those wellformed parameter names while creating options_mask
+                wellformed_params = gen_state["param_names"]
+
+                # Form the parameter name with the current sampled token id
+                if len(wellformed_params) == 0:
+                    curr_text = gen_state["curr_text"][
```

Comment:
Can we use curr_text with removing previous components, so we don't have to handle this. 
For example, when we enter state=function; curr_text = function_name in progress 
when we enter state=parameter --> curr_text=parameter_name in progress
when we enter state=parameter-value --> curr_text=parameter-value in progress

Reply:
Gotcha